core/src/main/scala/org/bitcoins/core/crypto/MnemonicCode.scala

```diff
@@ -0,0 +1,257 @@
+          search(start, mid - 1) // narrow the field
+        else search(mid + 1, end)
+      }
+
+      val indexOpt = search()
+
+      val index = indexOpt.getOrElse(
+        throw new IllegalArgumentException(s"Couldn't find $word in word list!")
+      )
+
+      BitVector.fromInt(index, size = MnemonicCode.BIT_GROUP_LENGTH)
+
+    }
+
+    val bits: BitVector = words
+      .map(curr => bitsForWord(curr))
+      .fold(BitVector.empty)((accum, curr) => accum ++ curr)
+
+    val codeInfo = MnemonicCode.getMnemonicCodeInfo(words)
+
+    bits.take(codeInfo.entropyAndChecksumBits)
+  }
+}
+
+/**
+  * @see [[https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki BIP39]]
+  */
+object MnemonicCode {
+
+  private[crypto] val BIT_GROUP_LENGTH = 11
+
+  private case class MnemonicCodeImpl(words: Vector[String])
+      extends MnemonicCode
+
+  /**
+    * Creates a mnemonic code from the provided words
+    */
+  def fromWords(words: Vector[String]): MnemonicCode = MnemonicCodeImpl(words)
+
+  /**
+    * Generates a mnemonic code from the given entropy bytes
+    */
+  def fromEntropy(entropy: ByteVector): MnemonicCode =
+    fromEntropy(entropy.bits)
+
+  /**
+    * The minimum number of entropy bits needed to construct a mnemonic code
+    */
+  private[crypto] val MIN_ENTROPY_BITS = 128
+
+  /**
+    * The maximum number of entropy bits allowed to construct mnemonic code
+    */
+  private[crypto] val MAX_ENTROPY_BITS = 256
+
+  /**
+    * The entropy for a mnemonic code has to be a multiple of 32
+    */
+  private[crypto] val ENTROPY_MULTIPLE = 32
+
+  /**
+    * Generates a mnemonic code from the given entropy bits
+    */
+  def fromEntropy(entropy: BitVector): MnemonicCode = {
+    require(
+      entropy.length >= MIN_ENTROPY_BITS,
+      s"Entropy must be at least $MIN_ENTROPY_BITS bits"
+    )
+
+    require(
+      entropy.length <= MAX_ENTROPY_BITS,
+      s"Entropy cannot be longer than $MAX_ENTROPY_BITS bits"
+    )
+
+    require(
+      entropy.length % ENTROPY_MULTIPLE == 0,
+      s"Entropy must be a multiple of $ENTROPY_MULTIPLE bits!"
+    )
+
+    val ENTROPY_CHECKSUM_DIVISOR = 32
+    val checksumLength = entropy.length / ENTROPY_CHECKSUM_DIVISOR
+
+    val hashedEntropy = CryptoUtil.sha256(entropy)
+    val hashedEntropyBits = hashedEntropy.bytes.toBitVector
+    val checkSum = hashedEntropyBits.take(checksumLength)
+    val entropyWithChecksum = entropy ++ checkSum
+
+    val bitGroups = entropyWithChecksum.grouped(BIT_GROUP_LENGTH)
+    val words = bitGroups.map { group =>
+      val index = group.toInt(signed = false)
+      ENGLISH_WORDS(index)
+    }.toVector
+
+    MnemonicCodeImpl(words)
+  }
+
+  /**
+    * Generates the specified bits of entropy
+    */
+  private def getEntropy(bits: Int): BitVector = {
+    val randomGenerator: SecureRandom = new SecureRandom
+
+    val byteArray: Array[Byte] = new Array[Byte](bits / 8)
+    randomGenerator.nextBytes(byteArray)
+    val bitVector = BitVector(byteArray)
```

Comment:
It might be wise to have some sort of invariant that the `bits % 8 == 0`. The ensuring clause will catch this invariant, but it will be baffling to the user why they did not get enough entropy.

Reply:
It is a  private method, though. I think that makes a `require` a bit overkill. If you disagree I can add it in